mongodbatlas/fw_resource_mongodbatlas_project.go

```diff
@@ -0,0 +1,857 @@
+		if err != nil {
+			return fmt.Errorf("error adding teams to the project: %v", err.Error())
+		}
+	}
+	return nil
+}
+
+func hasTeamsChanged(planTeams, stateTeams []tfTeamModel) bool {
+	sort.Slice(planTeams, func(i, j int) bool {
+		return planTeams[i].TeamID.ValueString() < planTeams[j].TeamID.ValueString()
+	})
+	sort.Slice(stateTeams, func(i, j int) bool {
+		return stateTeams[i].TeamID.ValueString() < stateTeams[j].TeamID.ValueString()
+	})
+	return !reflect.DeepEqual(planTeams, stateTeams)
+}
+
+func hasLimitsChanged(planLimits, stateLimits []tfLimitModel) bool {
+	sort.Slice(planLimits, func(i, j int) bool {
+		return planLimits[i].Name.ValueString() < planLimits[j].Name.ValueString()
+	})
+	sort.Slice(stateLimits, func(i, j int) bool {
+		return stateLimits[i].Name.ValueString() < stateLimits[j].Name.ValueString()
+	})
+	return !reflect.DeepEqual(planLimits, stateLimits)
+}
+
+func updateProject(ctx context.Context, conn *matlas.Client, projectState, projectPlan *tfProjectRSModel) error {
+	if projectPlan.Name.Equal(projectState.Name) {
+		return nil
+	}
+
+	projectID := projectState.ID.ValueString()
+
+	if _, _, err := conn.Projects.Update(ctx, projectID, newProjectUpdateRequest(projectPlan)); err != nil {
+		return fmt.Errorf("error updating the project(%s): %s", projectID, err)
+	}
+
+	return nil
+}
+
+func newProjectUpdateRequest(tfProject *tfProjectRSModel) *matlas.ProjectUpdateRequest {
+	return &matlas.ProjectUpdateRequest{
+		Name: tfProject.Name.ValueString(),
+	}
+}
+
+func deleteProject(ctx context.Context, conn *matlas.Client, projectID string) error {
+	stateConf := &retry.StateChangeConf{
+		Pending:    []string{"DELETING", "RETRY"},
+		Target:     []string{"IDLE"},
+		Refresh:    resourceProjectDependentsDeletingRefreshFunc(ctx, projectID, conn),
+		Timeout:    30 * time.Minute,
+		MinTimeout: 30 * time.Second,
+		Delay:      0,
+	}
+
+	_, err := stateConf.WaitForStateContext(ctx)
+
+	if err != nil {
+		tflog.Info(ctx, fmt.Sprintf("[ERROR] could not determine MongoDB project %s dependents status: %s", projectID, err.Error()))
+	}
+
+	_, err = conn.Projects.Delete(ctx, projectID)
+
+	return err
+}
+
+/*
+This assumes the project CRUD outcome will be the same for any non-zero number of dependents
+
+If all dependents are deleting, wait to try and delete
+Else consider the aggregate dependents idle.
+
+If we get a defined error response, return that right away
+Else retry
+*/
+func resourceProjectDependentsDeletingRefreshFunc(ctx context.Context, projectID string, client *matlas.Client) retry.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		var target *matlas.ErrorResponse
+		clusters, _, err := client.AdvancedClusters.List(ctx, projectID, nil)
+		dependents := AtlastProjectDependents{AdvancedClusters: clusters}
+
+		if errors.As(err, &target) {
+			return nil, "", err
+		} else if err != nil {
+			return nil, "RETRY", nil
```

Comment:
This is called when deleting project. RETRY is returned so it can be checked as a status for a refresh function. See https://github.com/mongodb/terraform-provider-mongodbatlas/pull/1365/files#diff-c2af32c41f07d77e630b76e890c8a1dfee1e053fdeac86f8fd31a0bb8e5c9663R801

Reply:
updated to constant.